### ISSUE TITLE
Saliency or Attention Visualization of Text Regression Transformer models

### ISSUE THREAD

**Issue by sachn1** [amaiya/ktrain#299]
Is it possible to get in and tackle with the gradients of say, a bert model, to visualize the saliency heatmap or visualize the attention weights of the heads? I know that currently LIME visualizations are not supported for text regression tasks within the library. 

I am currently working on a text regression task and I was looking out ways for its explainability. Are there some ways to get around with this? I think if we had access to the trained tokenizers along with the model, it should work as shown in [bertviz](https://github.com/jessevig/bertviz/blob/master/head_view_distilbert.ipynb). But upon saving a ktrain model what we could get is the model and preproc file right?

**Comment by amaiya** [comment]
First, **ktrain** is just a lightweight wrapper, so nothing in **ktrain** prevents you from accessing the underlying libraries like **transformers**.    Note also that there are two way to use BERT (proper) in **ktrain**: `keras_bert` and `transformers`.  I am assuming you're referring to `transformers` (e.g., using [Transformer API](https://nbviewer.jupyter.org/github/amaiya/ktrain/blob/master/tutorials/tutorial-A3-hugging_face_transformers.ipynb) in **ktrain**.).  For DistilBERT, `transformers` is always used.

Second, the tokenizers are not trained in pretrained models. You can re-create them directly with **transformers** (or `keras_bert`) if you need them.  Or you can access them with the `preproc` object.

```python

tokenizer = predictor.preproc.get_tokenizer() # works for transformers models like DistilBERT
```
In a nutshell, yes, you should be able to easily use `bertviz` with **ktrain**.

P.S. If using `keras_bert` (i.e., you created model specifically with `text_regression_model('bert', ...)` , you'll need to temporarily access tokenizer with `preproc.tok` instead of `preproc.get_tokenizer()`.  This'll be changed/fixed later.

**Comment by sachn1** [comment]
Thanks @amaiya! Just one last question - how do we access the attention weights? Or how do we set `output_attention = True` to the configuration?

**Comment by amaiya** [comment]
The underlying `transformers` model is directly accessible with `learner.model` or `predictor.model`.  In fact, you can load and use the model **outside** of *ktrain* with `TFAutoModelForSequenceClassification.from_pretrained`.  So, you can generate the visualization how ever `bertviz`  or `transformers` recommends to do it.  I would just follow the notebook you referenced earlier to generate the visualization.  If you have trouble, you can load the model as a PyTorch model with `from_pt=True` argument to `from_pretrained`.

**Comment by sachn1** [comment]
Thanks a lot @amaiya! Got it all working! I spent a lot of time to understand how to access the methods or configs of the underlying libraries. But posting the question here saved me a lot of time for sure!
